### ISSUE TITLE
The .net codegen path ignores a set_param override of a derived parameter: the sensitivity RHS keeps the stale chain rule even on a cold cache, and the .so cache key lets overridden and pristine models be served each other's sensitivity code

### ISSUE THREAD

**Issue by wshlavacek** [lanl/bngsim#694]
## Summary

Since #188, writing a **derived** parameter with `set_param` overrides it. This is BNG's `setParameter` semantics. The `Model.set_param` docstring states what that does to sensitivities (`_model.py:1096-1101`): a derived parameter pinned to a literal "no longer carries the chain rule from the primaries underneath it, so those primaries lose that reaction's term from their sensitivity columns, and the generated RHS changes to match."

The model-based codegen path (SBML, Antimony) does this. The `.net` codegen path does not. That is the path the `Simulator` prefers for every `Model.from_net` and `Model.from_bngl` model (`_simulator.py:1871-1876`). There are two failures, and both come from the `.net` path never reading the model's live attachment vector (`param_is_expression` / `codegen_data()["parameters"][i]["is_const"]`):

1. **An Elementary rate keeps the stale chain rule, even on a cold cache.** `generate_sens_rhs_c` decides which parameters are primary and which are derived from the `.net` text. The text still says `k2 = 2*k1`, so the code emits `dk2/dk1 * df/dk2` into `k1`'s column after `k2` has been pinned.
2. **A Functional rate is right on its first build, but the cache key cannot tell the two states apart.** Its sensitivity code comes from the live model (the GH #67 hook). But the `.so` is keyed on the `.net` bytes plus emit flags only. Whichever attachment state compiles first is then served to the other: in later processes that share `BNGSIM_CODEGEN_CACHE_DIR`, and within one process through the `prepare_codegen` memo. An override run can leave a **pristine** model with a zero sensitivity column.

Nothing warns. `param_is_expression` reports the override correctly, the trajectory is correct, and the wrong column has a plausible magnitude. Only sensitivities and everything built on them are affected: gradients, the FIM, and chunked JAX gradients. SBML and Antimony models are not affected.

Two model files, each with one reaction `A -> B` and `A(0) = 10`, `k1 = 1`, `k2 = 2*k1`. The override is `set_param("k2", 5.0)`. Each case was run in a fresh process:

| Rate law | Run | Cache | `dA/dk1` at t = 1 | Closed form |
|---|---|---|---|---|
| Elementary, `k2` | override | empty | **-0.134759** | 0 |
| Functional, `k2*Atot` | override, first | empty | 0 | 0 |
| Functional, `k2*Atot` | pristine, after that | hit | **0** | -0.453515 |
| Functional, `k2*Atot` | pristine, first | empty | -0.453515 | -0.453515 |
| Functional, `k2*Atot` | override, after that | hit | **-0.076894** | 0 |

The closed forms are `A = 10 exp(-k2 t)` (Elementary) and `A = 10 / (1 + 10 k2 t)` (Functional, `dA/dt = -k2 A^2`). When `k2` is pinned, `dA/dk1 = 0`. When `k2 = 2 k1`, `dA/dk1 = 2 dA/dk2`. Each wrong value is exactly the stale term: -0.134759 = -2·t·A(1) at `k2 = 5`, and -0.076894 = -200/51².

The same defect reaches two more public entry points:

- **`Model.from_bngl`.** Rules `A() -> C() kf` and `C() -> A() kd`, with `kd kcat*2` and `kd` overridden to 5. `dX(10)/dkcat` comes out as `[0.23496, -0.23496]`. A central FD of the trajectory gives `[0, 0]`.
- **`bngsim.jax.differentiable_solve(..., flat=True, chunk_size=2)`.** Flat mode pins every derived parameter with `force_override` (`_jax_bridge.py:404-405`). The chunked branch then builds a `Simulator` on the clone (`_jax_bridge.py:408`), which takes the `.net` path. On `tests/data/derived_rate_const.net`, `grad[kon]` is `0.0453515` (= `chi * grad[_rateLaw1]`) at `chunk_size=2`. It is `0` at `chunk_size=0`, and `0` is what `test_flat_grad_kon_is_zero` (`python/tests/test_jax_bridge.py:848`) asserts.

Measured at 3f4a76b; the cited lines are unchanged at 5294790 (and at 4ebeeb6).

## Reproduction

```bash
mkdir -p /tmp/netoverride && cd /tmp/netoverride

cat > elem.net <<'NET'
begin parameters
    1 k1   1     # Constant
    2 k2   2*k1  # ConstantExpression
end parameters
begin species
    1 A() 10
    2 B() 0
end species
begin reactions
    1 1 2 k2
end reactions
NET

cat > func.net <<'NET'
begin parameters
    1 k1   1     # Constant
    2 k2   2*k1  # ConstantExpression
end parameters
begin species
    1 A() 10
    2 B() 0
end species
begin groups
    1 Atot 1
end groups
begin functions
    1 frate() k2*Atot
end functions
begin reactions
    1 1 2 frate
end reactions
NET

cat > run.py <<'PY'
import sys, bngsim
path, mode = sys.argv[1], sys.argv[2]
m = bngsim.Model.from_net(path)
if mode == "override":
    m.set_param("k2", 5.0)   # BNG setParameter semantics: pin k2, detach it from k1
attached = dict(zip(m.param_names, m.param_is_expression))["k2"]
sim = bngsim.Simulator(m, method="ode", sensitivity_params=["k1"])
r = sim.run(t_span=(0.0, 1.0), n_points=11, rtol=1e-10, atol=1e-12)
print(f"{path} {mode:8s} k2_attached={attached!s:5s} cache_hit={sim.codegen_cache_hit!s:5s} "
      f"A(1)={r.species[-1, 0]:.6f}  dA/dk1(1)={r.sensitivities[-1, 0, 0]:+.6f}")
PY

echo "# 1. Elementary rate, override, empty cache"
export BNGSIM_CODEGEN_CACHE_DIR=$(mktemp -d)
python run.py elem.net override

echo "# 2. Functional rate, one cache dir, two processes: override first"
export BNGSIM_CODEGEN_CACHE_DIR=$(mktemp -d)
python run.py func.net override
python run.py func.net pristine

echo "# 3. Same, pristine first"
export BNGSIM_CODEGEN_CACHE_DIR=$(mktemp -d)
python run.py func.net pristine
python run.py func.net override
```

Output:

```
# 1. Elementary rate, override, empty cache
elem.net override k2_attached=False cache_hit=False A(1)=0.067379  dA/dk1(1)=-0.134759
# 2. Functional rate, one cache dir, two processes: override first
func.net override k2_attached=False cache_hit=False A(1)=0.196078  dA/dk1(1)=+0.000000
func.net pristine k2_attached=True  cache_hit=True  A(1)=0.476190  dA/dk1(1)=+0.000000
# 3. Same, pristine first
func.net pristine k2_attached=True  cache_hit=False A(1)=0.476190  dA/dk1(1)=-0.453515
func.net override k2_attached=False cache_hit=True  A(1)=0.196078  dA/dk1(1)=-0.076894
```

Expected, from the closed forms above: case 1 gives `dA/dk1 = 0`; case 2 gives 0, then -0.453515; case 3 gives -0.453515, then 0. Every `A(1)` is correct (`10e^-5`, `10/51`, `10/21`). A central FD of `A(1)` in `k1` through plain runs on `elem.net`, with the override held, also gives `0.0`.

The same models on the model-based path honour the override:

```python
import bngsim
ANT = """
model m
  compartment c = 1
  species A in c = 10, B in c = 0
  k1 = 1
  k2 = 2*k1
  R1: A -> B; k2*A{extra}
end
"""
for name, extra in (("elementary", ""), ("functional", "*A")):
    m = bngsim.Model.from_antimony_string(ANT.format(extra=extra))
    m.set_param("k2", 5.0)
    r = bngsim.Simulator(m, method="ode", sensitivity_params=["k1"]).run(
        t_span=(0.0, 1.0), n_points=11, rtol=1e-10, atol=1e-12)
    print(f"SBML path, {name}, override: dA/dk1(1)={r.sensitivities[-1, 0, 0]:+.6f}")
# SBML path, elementary, override: dA/dk1(1)=+0.000000
# SBML path, functional, override: dA/dk1(1)=+0.000000
```

## Cause

Two paired sites classify parameters as primary or derived, and only one of them reads the live model:

- **Model path.** `generate_sens_from_model` takes `primary_param_names` and `derived_exprs` from `codegen_data()`'s live `is_const` (`python/bngsim/_codegen.py:9797-9801`).
- **`.net` path.** `generate_sens_rhs_c(net_path)` re-parses the file and takes them from the text (`_codegen.py:3813-3814`). The `is_const` there comes from `_classify_parameter_kinds` (`_codegen.py:1013`), which reads the file's expressions. The function takes no model argument, so a live override cannot reach it. Its call in `generate_combined_c` (`_codegen.py:4941-4943`) passes only the path. **This is failure 1.**

For a Functional model, `generate_sens_rhs_c` declines. `generate_combined_c` then appends `generate_sens_from_model(model, functional=True)` (`_codegen.py:4945-4957`), plus the model-derived Jacobian, outputs and output sensitivities. All of these read the live model, so the source does depend on the attachment state. The keys it is cached under do not:

- The in-process memo key (`_codegen.py:11727-11736`) is the `.net` path, the emit flags, the budget tag and the chunk policy.
- The on-disk key is `compute_model_hash(net_path)` (`_codegen.py:4984-5020`, the file bytes plus `.tfun` dependencies) plus the flag suffix (`_codegen.py:11760-11798`).

Neither key contains the attachment vector. **This is failure 2.** Once failure 1 is fixed, the Elementary source will depend on the attachment state too, so the key has to change either way.

The model path already solved the key half. `compute_model_codegen_hash` hashes `codegen_data()` (`_codegen.py:11124`, `11143`) precisely because it "carries the attachment vector `is_const`, which is live state a `set_param` on a derived parameter moves (issue #188) and which the source genuinely depends on" (`_codegen.py:11097-11100`). It also folds in `switch_gate_cache_digest` (`_codegen.py:11166`). The `.net` key got neither. The #188 tests (`python/tests/test_derived_param_write.py`) load only SBML through `Model.load`, so the `.net` path has no coverage.

## Fix

- **Classification.** Give `generate_sens_rhs_c` the live attachment vector when a model is supplied. `generate_combined_c` has the model in hand. Take `is_const` from `model.param_is_expression` / `codegen_data()`, as `generate_sens_from_model` does at `:9797`, instead of from the parsed text. With `model=None` the function should keep today's text-derived classification, which is correct for a pristine file. `prepare_codegen_source` (`_codegen.py:11429`, the MIR backend's `.net` entry) goes through the same `generate_combined_c` call, so it picks this up too. I did not measure the MIR backend, because this build lacks it.
- **Keys.** Fold the live attachment vector into both the `prepare_codegen` memo key (`:11727`) and the on-disk suffix (`:11760-11798`). Add it only when it differs from the file's own classification, following the convention of the `:chunk=` and `:no_functional_sens` suffixes, so every `.so` already cached for a pristine file keeps its key. Check whether the `.net` key also needs `switch_gate_cache_digest`. The model path includes it because `generate_sens_from_model` reads it, and the `.net` path appends that same code.
- **Tests.** Add regression tests through `Model.from_net` and `Model.from_bngl` with a derived-parameter override, covering: an Elementary cold build; a Functional model with both cross-process orders against one cache dir, plus the in-process memo; and `differentiable_solve(..., flat=True, chunk_size=2)` against `test_flat_grad_kon_is_zero`'s contract.

## Related

- #188: the override semantics this path does not honour. Its fix covered the model path only.
- #174: the structural `.so` key. It added `is_const` to the model-path key only.
- #181: the `.net` text classification that `generate_sens_rhs_c` uses.
- #67: the Functional sensitivity hook that appends live-model code under the file-only key.
- #15: why the `Simulator` prefers the `.net` path for `from_net`/`from_bngl` models.

**Comment by wshlavacek** [comment]
Part of a family of related defects tracked in #784. Siblings: #689, #699, #721, #730, #731, #734.
